Dan/solution.py

```diff
@@ -0,0 +1,119 @@
+import random
+
+RIGHT = 'r'
+DOWN = 'd'
+
+class Grid:
+    """
+    I created a class so I could reference the grid without using global variables
+    """
+    def __init__(self,m,n):
+        """
+        creates a playing board for the robot
+        """
+        self.columns = m
+        self.rows = n
+        self.board = makeBoard(m,n)
+
+    def find_good_paths(self):
+        """
+        Returns an array of all the valid paths through the board
+        """
+        return self.robot_step((0,0),[])
+
+    def robot_step(self, location, current_path, direction=None):
+        """
+        Robot tries to move in the direction specifid. If the robot can move in
+        that direction it adds the direction to the path(which is the current
+        steps that it has taken up to the current location). If the new space is
+        the bottom right corner then it returns the path since it is a complete
+        valid path. If it is not the finish we check the valid paths if the
+        robot moves right or down from this location.
+
+        If the robot cannot take the step then this returns None.
+        """
+        new_location = self.take_step(location, direction)
+        if new_location:
+            if new_location != location:
+                current_path.append(direction)
+            if self.is_finish(new_location):
+                return [current_path]
+            right = self.robot_step(new_location,list(current_path),RIGHT)
+            down = self.robot_step(new_location,list(current_path),DOWN)
+            if right and down:
+                return right + down
+            elif right:
+                return right
+            elif down:
+                return down
+            else:
+                return None
+        else:
+            return None
+
+    def take_step(self, location, direction):
+        """
+        Based on the location passed in and the direction this returns the new
+        location. If the new location is not valid it returns None
+
+        If no direction is specified then it returns the current location (this)
+        is for the first space)
+        """
+        if direction == RIGHT:
+            step_location = (location[0]+1,location[1])
```

Comment:
This step is not incorrect, but the general tendency is to read as (row, column). My first thought was why are u increasing the row number if you're moving right, but later I realized that [0] is the column in your case.

Reply:
Oh really? That's good to know. I was just thinking about mathematical coordinates which are written (x,y)

Reply:
Ah makes sense... I was thinking about matrix since the beginning and maybe that clouded my judgement